Add fetch calls

npm1 opened the pull request:
Relevant issues:
* https://github.com/fedidcg/FedCM/issues/261
* https://github.com/fedidcg/FedCM/issues/320

This PR adds the fetch calls to the other FedCM fetches. It is pretty noncontroversial but is missing the majority of request parameters. A future PR will address that.


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/npm1/FedCM/pull/356.html" title="Last updated on Sep 30, 2022, 7:39 PM UTC (b930834)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/fedidcg/FedCM/356/1572b7a...npm1:b930834.html" title="Last updated on Sep 30, 2022, 7:39 PM UTC (b930834)">Diff</a>

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,47 @@ To <dfn>compute account state</dfn> given an {{IdentityProvider}} |provider| and
+                |account|["email"].
+            1. If |account|["given_name"] exists and is a [=string=], set |acc|'s
+                {{account_json/given_name}} to |account|["given_name"].
+            1. If |account|["approved_clients"] exists and is a list where each item is a
```
maybe approved_clients need to default to empty list by default (otherwise you are going to have to test for its presence later?)? i guess either way could work.

npm1 replied:
It is optional.

samuelgoto replied:
sgtm

samuelgoto commented on spec/index.bs:
```diff
@@ -1222,13 +1251,30 @@ To <dfn>request consent to sign-up</dfn> the user with a given [=Account JSON=]
+        |manifest|["{{Manifest/client_metadata_endpoint}}"] (the relative URL) and |provider|'s
+        {{IdentityProvider/configURL}} (the base URL).
+    1. If |clientMetadataUrl| is failure, return null.
+    1. Let |request| be a new <a spec=fetch for=/>request</a> whose [=request/url=] is
```
It is important that we capture here the fact in line 1129 that you are removing that NO cookies are passed to this request.

samuelgoto commented on spec/index.bs:
```diff
@@ -1248,16 +1294,36 @@ To <dfn>sign-in</dfn> the user with a given an [=AccountState=] |accountState|:
+    1. Let |request| be a new <a spec=fetch for=/>request</a> whose [=request/url=] is
+        |tokenUrl|, [=request/redirect mode=] set to "error", [=request/mode=] set to `POST`, and
+        [=request/body=] set to |requestBody|.
+    1. TODO: set other fields of the request, including the [[#Sec-FedCM-CSRF|Sec-FedCM-CSRF]]
```
Again, here goes the user's IdP's cookies.

samuelgoto reviewed (commented):
Looks generally good to me, but we need to resolve TODOs before merging. Will take another pass when you resolve them.

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,47 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
+    1. If |accountsUrl| is failure, return an empty list.
+    1. Let |request| be a new <a spec=fetch for=/>request</a> whose [=request/url=] is
+        |accountsUrl| and [=request/redirect mode=] set to "error".
+    1. TODO: set other fields of the request, including the [[#Sec-FedCM-CSRF|Sec-FedCM-CSRF]]
```
For all of these requests, can you make sure that we set the flags such that:

- we don't follow redirects and
- we don't get intercepted by service works and
- we fail when it is not HTTPs?

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,47 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
-        [[RFC9110#header.location|HTTP redirects]] and instead return null if there is any error.
-        See also [[#idp-api-accounts-endpoint]].
+    1. Let |accountsUrl| be the result of running [=url parser=] given
+        |manifest|[{{Manifest/accounts_endpoint}}] (the relative URL) and |provider|'s
```
Can we make sure we check that all of the endpoints in the |manifest| are same-origin with the |manifest| itself?

npm1 commented:
PTAL, addressed all of your comments. The pending thing is the referrer being handwavy (in the case we do want to send it), while I figure out how to set it properly.

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,58 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
-        See also [[#idp-api-accounts-endpoint]].
+    1. Let |accountsUrl| be the result of running [=url parser=] given
+        |manifest|[{{Manifest/accounts_endpoint}}] (the relative URL) and |provider|'s
+        {{IdentityProviderConfig/configURL}} (the base URL).
```
Don't you need to check that configURL and accountsURL is same-origin? Here and elsewhere?

samuelgoto replied:
ah, nevermind, i see line #1189 now.

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,58 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
-        [[RFC9110#header.location|HTTP redirects]] and instead return null if there is any error.
-        See also [[#idp-api-accounts-endpoint]].
+    1. Let |accountsUrl| be the result of running [=url parser=] given
+        |manifest|[{{Manifest/accounts_endpoint}}] (the relative URL) and |provider|'s
```
This isn't guaranteed to be a "relative URL" right? What happens if an absolute URL is passed?

npm1 replied:
Done

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,58 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
+        1. [=request/client=] set to null
+        1. [=request/window=] set to "no-window"
+        1. [=request/service-workers mode=] set to "none"
+        1. [=request/destination=] set to "identity"
```
```suggestion
        1. [=request/destination=] set to "web-identity"
```

"identity" is probably too broad/ambiguous (e.g. maybe it means "cookies"?)?

npm1 replied:
Cookies are a web thing too? Unclear how you can have a non-web destination in fetch, but sure I don't have strong opinions about the name.

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,58 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
+        1. [=request/window=] set to "no-window"
+        1. [=request/service-workers mode=] set to "none"
+        1. [=request/destination=] set to "identity"
+        1. [=request/origin=] set to a unique [=opaque origin=]
```
what does this do? is this for the cache partitioning?

npm1 replied:
It sets the origin of the request. The origin is 'user agent' but there is not such thing.

samuelgoto replied:
Is this supposed to be referring to this [1]? [=opaque origin=] serializes to "null" rather than "user agent" right [2]?

Do you know what this is for?

[1] https://fetch.spec.whatwg.org/#concept-request-origin
[2] https://html.spec.whatwg.org/multipage/origin.html#concept-origin-opaque

samuelgoto replied:
If we don't know what this is for, why are we setting it? Is that a required field? Can we just leave it unset for now?

npm1 replied:
I do not think so, because Fetch does not handle that properly. `If request’s origin is "client", then set request’s origin to request’s client’s origin.`
https://fetch.spec.whatwg.org/#ref-for-concept-request-origin%E2%91%A6:~:text=If%20request%E2%80%99s%20origin%20is%20%22client%22%2C%20then%20set%20request%E2%80%99s%20origin%20to%20request%E2%80%99s%20client%E2%80%99s%20origin.

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,58 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
-        See also [[#idp-api-accounts-endpoint]].
+    1. Let |accountsUrl| be the result of running [=url parser=] given
+        |manifest|[{{Manifest/accounts_endpoint}}] (the relative URL) and |provider|'s
+        {{IdentityProviderConfig/configURL}} (the base URL).
```
IdentityProviderConfig/configURL is also a "string". don't you need to parse it?

npm1 replied:
Done (it was handwavy everywhere)

samuelgoto commented on spec/index.bs:
```diff
@@ -1222,13 +1262,41 @@ To <dfn>request consent to sign-up</dfn> the user with a given [=Account JSON=]
+        1. [=request/client=] set to null
+        1. [=request/window=] set to "no-window"
+        1. [=request/service-workers mode=] set to "none"
+        1. [=request/destination=] set to "identity"
```
```suggestion
        1. [=request/destination=] set to "web-identity"
```

Same here, I think something narrower would be useful.

samuelgoto commented on spec/index.bs:
```diff
@@ -1248,16 +1316,48 @@ To <dfn>sign-in</dfn> the user with a given an [=AccountState=] |accountState|:
+    1. If |tokenUrl| is not [=same origin=] with |provider|'s {{IdentityProviderConfig/configURL}},
+        return null.
+    1. If |tokenUrl| is not a [=potentially trustworthy URL=], return null.
+    1. Let |requestBody| be a new object constructed as follows (TODO: what is the correct type):
```
Would that be `application/x-www-form-urlencoded`?

npm1 replied:
Not sure how to set the body for application/x-www-form-urlencoded, need to ask.

samuelgoto commented on spec/index.bs:
```diff
@@ -1248,16 +1316,48 @@ To <dfn>sign-in</dfn> the user with a given an [=AccountState=] |accountState|:
-    1. Let |tokens| be the result of making a POST request described in
-        [[#idp-api-id-token-endpoint]].
-    1. Return |tokens|["token"].
+    1. Let |tokenUrl| be the result of running [=url parser=] given
```
```suggestion
    1. Let |idTokenUrl| be the result of running [=url parser=] given
```

For consistency?

samuelgoto commented on spec/index.bs:
```diff
@@ -1248,16 +1316,48 @@ To <dfn>sign-in</dfn> the user with a given an [=AccountState=] |accountState|:
+        1. [=request/client=] set to null
+        1. [=request/window=] set to "no-window"
+        1. [=request/service-workers mode=] set to "none"
+        1. [=request/destination=] set to "identity"
```
```suggestion
        1. [=request/destination=] set to "web-identity"
```

samuelgoto replied:
Still unresolved?

npm1 replied:
I'm confused. What's unresolved? Are you looking at the latest version?

samuelgoto commented on spec/index.bs:
```diff
@@ -1248,16 +1316,48 @@ To <dfn>sign-in</dfn> the user with a given an [=AccountState=] |accountState|:
+    1. Let |token| be null.
+    1. [=Fetch=] |request| with <a spec=fetch>processResponseConsumeBody</a> set to the following
+        steps given a <a spec=fetch for=/>response</a> |response|:
+        1. If |response| is a [=network error=] or its [=response/status=] is not an [=ok status=],
```
This block shows up multiple times .. I'm wondering if we should create an algorithm and reuse it in the other algorithms. WDYT?

npm1 replied:
Done

samuelgoto commented on spec/index.bs:
```diff
@@ -1280,13 +1380,24 @@ manipulation to fingerprint. See [[#manifest-fingerprinting]].
+    1. If |rootUrl| is not a [=potentially trustworthy URL=], return false.
+    1. Let |request| be a new <a spec=fetch for=/>request</a> with the following properties:
+        1. [=request/url=] set to |rootUrl|
+        1. [=request/redirect mode=] set to "error"
```
I think that following redirects here is OK ... e.g. it does currently redirect from "https://google.com/.well-known/web-identity" to "https://www.google.com/.well-known/web-identity", I think.

npm1 replied:
Done

samuelgoto replied:
still unresolved?

samuelgoto commented on spec/index.bs:
```diff
@@ -1309,18 +1420,48 @@ whether the manifest is included in the manifest list:
+    1. If |url| is not a [=potentially trustworthy URL=], return null.
+    1. Let |request| be a new <a spec=fetch for=/>request</a> with the following properties:
+        1. [=request/url=] set to |url|
+        1. [=request/redirect mode=] set to "error"
```
I'm also wondering if following redirects here is also OK.

samuelgoto replied:
Also, the client_metadata_endpoint too, right?

npm1 replied:
I think impl only allows the root one to redirect

samuelgoto replied:
Yeah, only the root one. That works for me.

samuelgoto commented on spec/index.bs:
```diff
@@ -1309,18 +1420,48 @@ whether the manifest is included in the manifest list:
+        1. [=request/client=] set to null
+        1. [=request/window=] set to "no-window"
+        1. [=request/service-workers mode=] set to "none"
+        1. [=request/destination=] set to "identity"
```
```suggestion
        1. [=request/destination=] set to "web-identity"
```

npm1 commented:
Ready for more review, PTAL.

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1182,58 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
+        {{IdentityProviderConfig/configURL}} (the base URL).
+    1. If |accountsUrl| is failure, return an empty list.
+    1. If |accountsUrl| is not [=same origin=] with |provider|'s
+        {{IdentityProviderConfig/configURL}}, return an empty list.
```
I meant here in this line, configURL is a string, so you can't test if it is [=same origin=] with |accountsUrl| unless you [=url parser=]?

Here and elsewhere where you are using configURL in a way that is not a string.

samuelgoto commented on spec/index.bs:
```diff
@@ -1222,13 +1262,41 @@ To <dfn>request consent to sign-up</dfn> the user with a given [=Account JSON=]
+        |manifest|["{{Manifest/client_metadata_endpoint}}"] (the relative URL) and |provider|'s
+        {{IdentityProviderConfig/configURL}} (the base URL).
+    1. If |clientMetadataUrl| is failure, return null.
+    1. If |clientMetadataUrl| is not a [=potentially trustworthy URL=], return null.
```
ah, neat, this captures "localhost": https://w3c.github.io/webappsec-secure-contexts/#potentially-trustworthy-origin

samuelgoto commented on spec/index.bs:
```diff
@@ -1248,16 +1316,48 @@ To <dfn>sign-in</dfn> the user with a given an [=AccountState=] |accountState|:
+    1. Let |request| be a new <a spec=fetch for=/>request</a> with the following properties:
+        1. [=request/url=] set to |tokenUrl|
+        1. [=request/mode=] set to "POST"
+        1. [=request/body=] set to |requestBody|
```
This seems a bit off to me, because it seems that a [=request/body=] has to be one of the following: a stream, a source or a length [1], and here |requestBody| is neither of those.

https://fetch.spec.whatwg.org/#concept-body
https://fetch.spec.whatwg.org/#concept-request-body

samuelgoto commented on spec/index.bs:
```diff
@@ -1280,13 +1380,24 @@ manipulation to fingerprint. See [[#manifest-fingerprinting]].
+        1. [=request/client=] set to null
+        1. [=request/window=] set to "no-window"
+        1. [=request/service-workers mode=] set to "none"
+        1. [=request/destination=] set to "identity"
```
```suggestion
        1. [=request/destination=] set to "web-identity"
```

samuelgoto commented on spec/index.bs:
```diff
@@ -1248,16 +1316,48 @@ To <dfn>sign-in</dfn> the user with a given an [=AccountState=] |accountState|:
+        1. [=request/destination=] set to "identity"
+        1. [=request/origin=] set to a unique [=opaque origin=]
+        1. [=request/header list=] set to a [=list=] containing a single [=header=] with
+            [=header/name=] set to "Accept" and [=header/value=] set to "JSON"
```
```suggestion
            [=header/name=] set to "Accept" and [=header/value=] set to [=JSON MIME Type=]
```

Shouldn't this be a mime type?

samuelgoto reviewed (commented):
Looks generally good to me, but still feels like the parameters aren't quite right.

On a separate note, any chance we can get @domfarolino to take a pass at this too?

npm1 commented:
> Looks generally good to me, but still feels like the parameters aren't quite right.
> 
> On a separate note, any chance we can get @domfarolino to take a pass at this too?

Can you look again but looking at the latest version??

npm1 commented:
Added `Sec-` header changes

samuelgoto commented on spec/index.bs:
```diff
@@ -1181,19 +1183,72 @@ To <dfn>compute account state</dfn> given an {{IdentityProviderConfig}} |provide
+        1. [=request/destination=] set to "web-identity"
+        1. [=request/origin=] set to a unique [=opaque origin=]
+        1. [=request/header list=] set to a [=list=] containing a single [=header=] with
+            [=header/name=] set to "Accept" and [=header/value=] set to "JSON"
```
Again, shouldn't this be a mime type? Are you sure this isn't supposed to be text/json?

npm1 replied:
Again, don't think so? MIME type is an object computed from headers[1] but header names and values are byte sequences [2]. I think it should be `application/json` based on the implementation.
[1] https://mimesniff.spec.whatwg.org/#mime-type
[2] https://fetch.spec.whatwg.org/#concept-header

samuelgoto commented on spec/index.bs:
```diff
@@ -1222,13 +1277,41 @@ To <dfn>request consent to sign-up</dfn> the user with a given [=Account JSON=]
+    1. If |clientMetadataUrl| is failure, let |clientMetadataUrl| be the result of running
+        [=url parser=] given |manifest|["{{Manifest/client_metadata_endpoint}}"] (the relative URL) and |configUrl| (the base URL).
+    
+    Note: This means the we allow passing the client metadata endpoint as either an absolute or
```
This block is also going to be repeated 5 times (for each of the endpoints), so I think it is worth creating an algorithm to build these URLs form the baseURL (and check for things like same-site-ness).

npm1 replied:
It is 3 times, but yea worth doing the helper. Done

npm1 reviewed (commented):
Thanks for the review, PTAL.

samuelgoto commented:
LGTM

I'll merge, and if there is anything pending we can patch separately.
